### PR TITLE
os/bluestore: introduce multithireading sync for bluestore's repairer

### PR DESCRIPTION
**Pull request by ifed01**
In quick-fix mode bluestore uses 2 threads by default to perform the  repair. Due to lacking synchronization they might corrupt repair  transaction batch.

Fixes: https://tracker.ceph.com/issues/50017

Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by aclamk** [comment]
**this is unrelated to commited code**

In some cases, like `_fsck_check_object_omap()`, we put `_record_onode()` into rocksdb transaction and submit it immediately.
In other cases, like `repairer->fix_spanning_blobs()` we put `_record_onode()` into transaction to submit later.
```
    if (broken) {
      derr << "fsck error: " << oid << " - " << broken
           << " zombie spanning blob(s) found, the first one: "
           << *first_broken << dendl;
      if(repairer) {
        auto txn = repairer->fix_spanning_blobs(db);
	_record_onode(o, txn);
      }
    }
  }

  if (o->onode.has_omap()) {
    _fsck_check_object_omap(depth, o, ctx);
  }
  ```
I read it, that it is possible to:
1) fix spanning blobs in onode o, but not commit to DB yet
2) actually commit to DB even more modification of o - regarding omap handling
3) then commit to DB changes from step 1), forgetting fixes from 2)

Am I missing something?
Maybe we should always commit repair steps, and not aggregate them into huge transaction to execute at the end?

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -3645,13 +3648,32 @@ class BlueStoreRepairer
     ++to_repair_cnt;
   }
 
-  StoreSpaceTracker& get_space_usage_tracker() {
-    return space_usage_tracker;
+  void init_space_usage_tracker(
+    uint64_t total_space, uint64_t lres_tracking_unit_size)
+  {
+    //NB: not for use in multithreading mode!!!
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.h:3654]: "// Does not need lock - illegal to call in multithreading fsck" ?

File: src/test/objectstore/store_test.cc [hunk 2]
```diff
@@ -8339,6 +8339,76 @@ TEST_P(StoreTestSpecificAUSize, BluestoreRepairTest) {
 
 }
 
+TEST_P(StoreTestSpecificAUSize, BluestoreBrokenZombieRepairTest) {
```
Comment (aclamk) [inline src/test/objectstore/store_test.cc:8342]: +1 for test being in separate commit then fix.

**Review by aclamk — CHANGES_REQUESTED** [review]
Excellent unittest.
Maybe details to fix on implementation.

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -7524,8 +7524,11 @@ BlueStore::OnodeRef BlueStore::fsck_check_objects_shallow(
            << " zombie spanning blob(s) found, the first one: "
            << *first_broken << dendl;
       if(repairer) {
-        auto txn = repairer->fix_spanning_blobs(db);
-	_record_onode(o, txn);
+        repairer->fix_spanning_blobs(
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:7527]: The intention is to avoid private _record_onode() exposure - hence the design

**Comment by ifed01** [comment]
> **this is unrelated to commited code**
> 
> In some cases, like `_fsck_check_object_omap()`, we put `_record_onode()` into rocksdb transaction and submit it immediately.
> In other cases, like `repairer->fix_spanning_blobs()` we put `_record_onode()` into transaction to submit later.
> 
> ```
>     if (broken) {
>       derr << "fsck error: " << oid << " - " << broken
>            << " zombie spanning blob(s) found, the first one: "
>            << *first_broken << dendl;
>       if(repairer) {
>         auto txn = repairer->fix_spanning_blobs(db);
> 	_record_onode(o, txn);
>       }
>     }
>   }
> 
>   if (o->onode.has_omap()) {
>     _fsck_check_object_omap(depth, o, ctx);
>   }
> ```
> 
> I read it, that it is possible to:
> 
> 1. fix spanning blobs in onode o, but not commit to DB yet
> 2. actually commit to DB even more modification of o - regarding omap handling
> 3. then commit to DB changes from step 1), forgetting fixes from 2)
> 
> Am I missing something?
> Maybe we should always commit repair steps, and not aggregate them into huge transaction to execute at the end?

I agree that doesn't look elegant enough. Unfortunately it reflects the evolution (overburden) of fsck procedure - it supports  both real errors detection and naming scheme conversion for upgrade purposes. The former was originally designed to use a single repair transaction which looks safer from data consistency point of view. But the latter can't use all-in-one model due to potential data volume. Hence the mixture.
Regarding your concerns about overwriting previous modifications - perhaps you're right and it can fail this way. Will try to reproduce...

File: src/os/bluestore/BlueStore.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:7527]: Thx - I missed that!

**Review by aclamk — APPROVED** [review]

**Comment by ifed01** [comment]
> Regarding your concerns about overwriting previous modifications - perhaps you're right and it can fail this way. Will try to reproduce...
@aclamk - FYI: https://tracker.ceph.com/issues/50926
Good catch!

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-21_12:49:52-rados-wip-kefu-testing-2021-05-21-1851-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-22_15:42:09-rados-wip-kefu-testing-2021-05-22-2207-distro-basic-smithi/

failure tracked by 

- https://tracker.ceph.com/issues/50243
- https://tracker.ceph.com/issues/50151

**Pull request merged**